src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs

```diff
@@ -118,6 +116,23 @@ public virtual bool CanValidateToken
             get { return true; }
         }
 
+        private int CountTokenSegments(string token)
+        {
+            if (String.IsNullOrEmpty(token))
+                return 0;
+
+            // Counts number of Jwt segments, capped to MaxJwtSegmentCount + 1 to allow detecting when number of segments is too large.
+            // For example: "a.b.c.d.e.f.g.h" => [a], [b], [c], [d], [e], [f.g.h]. 6 segments.
+            // If just MaxJwtSegmentCount was used, then [a], [b], [c], [d], [e.f.g.h] would be returned. 5 segments.
+            int tokenPartsCount = 1;
+            for (int i = 0; i < token.Length && tokenPartsCount < JwtConstants.MaxJwtSegmentCount + 1; i++)
```

Comment:
You don't have to test `tokenPartsCount < JwtConstants.MaxJwtSegmentCount + 1` on every iteration.

Consider

            for (int i = 0; i < token.Length; i++)
            {
                if (token[i] == '.')
                {
                    tokenPartsCount++;
                    if (tokenPartsCount > JwtConstants.MaxJwtSegmentCount)
                        break;
                }
            }

Reply:
consider this solution as I think it is more efficient for this use-case:
```c#
        private static int CountSegments(string token)
        {
            if (string.IsNullOrEmpty(token))
                return 0;

            int tokenPartsCount = 1;
            int index = 0;

            while ((index = input.IndexOf('.', index)) != -1)
            {
                index++;
                tokenPartsCount++;
            }
            return tokenPartsCount;
        }
```

Reply:
That's not more efficient because of the extra function calls. IndexOf isn't able to avoid a linear scan.
You can't get much better than @KrisVendermotten's code -- although, I wouldn't be afraid to use `foreach(char ch in token)` here -- the JIT compiler does a stellar job of optimizing CharEnumerator. It's simple and compact.

But really, this is nit-picking over the main improvement here which is avoiding string.Split allocations.